File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        left = [1] * len(nums)
        right = [1] * len(nums)

        for i in range(1, n):
            left[i] = left[i-1] * nums[i-1]
            right[n-i-1] = right[n-i] * nums[n-i]
        
        res= []
        for i in range(n):
            res.append(left[i] * right[i]) 
        return res

File: Data Structures & Algorithms/sliding-window-maximum/submission-2.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> res;
        deque<int> dq;
        int l = 0;

        for (int i = 0; i < nums.size(); i++) {
            while (!dq.empty() && nums[dq.back()] <= nums[i]) {
                dq.pop_back();
            }

            dq.push_back(i);

            int window = i - l + 1;
            if (window == k) {
                res.push_back(nums[dq.front()]);
                if (dq.front() == l) {
                    dq.pop_front();
                }
                l++;
            }
        }

        return res;
    }
};
